browserdebuggertools/chrome/interface.py

```diff
@@ -185,9 +182,121 @@ def get_opened_javascript_dialog(self):
+        return self._get_outer_html(root_node_id)
+
+    def _get_iframe_html(self, xpath, _attempts=0):
+        # type: (str, int) -> str
+
+        try:
+            backend_node_id = self._get_iframe_backend_node_id(xpath)
+            return self._get_outer_html(backend_node_id)
+
+        except ResourceNotFoundError:
+            # The cached node doesn't exist any more, so we need to find a new one that matches
+            # the xpath. Backend node IDs are unique, so there is not a risk of getting the
+            # outer html of the wrong node.
+            if _attempts < 1:
+                if xpath in self._node_map:
+                    del self._node_map[xpath]
+                return self._get_iframe_html(xpath, _attempts=_attempts + 1)
+
+            raise IFrameNotFoundError()
+
+    def _get_iframe_backend_node_id(self, xpath):
+        # type: (str) -> int
+
+        if xpath in self._node_map:
+            return self._node_map[xpath]
+
+        try:
+            node_info = self._get_info_for_first_matching_node(xpath)
+            backend_node_id = node_info["node"]["contentDocument"]["backendNodeId"]
+        except (KeyError, ResourceNotFoundError):
+            raise IFrameNotFoundError()
+
+        self._node_map[xpath] = backend_node_id
+        return backend_node_id
+
+    def _get_info_for_first_matching_node(self, xpath):
+        # type: (str) -> dict
+
+        search_info = self._perform_search(xpath)
+        if search_info.get("resultCount", 0) > 0:
+            search_results = self._get_search_results(search_info["searchId"], 0, 1)
+            self._discard_search(search_info["searchId"])
+            return self._describe_node(search_results["nodeIds"].pop())
+        raise ResourceNotFoundError("No matching nodes for xpath: %s" % xpath)
+
+    def _perform_search(self, xpath):
+        # type: (str) -> dict
+
+        # DOM.getDocument must have been called on the current page first otherwise performSearch
+        # returns an array of 0s.
+        self._socket_handler.event_handlers["PageLoad"].check_page_load()
+        return self.execute("DOM", "performSearch", {"query": xpath})
+
+    def _get_search_results(self, search_id, from_index, to_index):
+        # type: (str, int, int) -> dict
+
+        return self.execute("DOM", "getSearchResults", {
+            "searchId": search_id, "fromIndex": from_index, "toIndex": to_index
+        })
+
+    def _discard_search(self, search_id):
+        # type: (str) -> None
+        """
+        Discards search results for the session with the given id. get_search_results should no
+        longer be called for that search.
+        """
+
+        self._socket_handler.execute("DOM", "discardSearchResults", {"searchId": search_id})
+
+    def _describe_node(self, node_id):
+        # type: (str) -> dict
+
+        return self._socket_handler.execute("DOM", "describeNode", {"nodeId": node_id})
```

Comment:
Should use `self.execute()` instead of `_socket_handler.execute()`

Reply:
This is irrelevant now because we moved this method onto the NodeManager